# Download the Google Street View Image

## Acquire a Google API Key

https://developers.google.com/maps/documentation/javascript/get-api-key

In [1]:
import urllib
import requests
import urllib.request
import pandas as pd
import time
import os
import re

In [2]:
# 3.1 Paste your Google API Key here, pay attention to the charge rules
## free for about 20,000 ~ 30,000 images
# -*- coding: utf-8 -*-

csv_dir = 'Columbia_SVIpoints_4326.csv'
image_dir = 'images_data' # image save path
key = 'AIzaSyBIsZXSsw4kbvOhlj6SXj1YIHeCCxtxnjw'  # your API
# YOUR_SIGNATURE = 'BASE64_SIGNATURE' # IF your API has signature, add this parameter


# In order to ensure that everyone downloads the photos with the same parameters, please do not change the following parameters
size =  '800x400'  # image size {width}x{height}
pitch= '0'        #(with 90 degrees indicating straight up,degrees -90 indicating straight down
fov = '90'

In [3]:

def download(url, save_dir,ID):
    try:
        conn = urllib.request.urlopen(url)
    except Exception as e:
        print('wrong',str(ID),e)
        return 0
    f = open(save_dir, 'wb')
    f.write(conn.read())
    f.close()
    print('image save', str(ID))
    return 0

## Read SVI sampling points

In [4]:
## # check the downloaded images
filepath_downloaded = []
for root, dirs, files in os.walk(image_dir ):
    for filename in files:
        if filename.endswith('.jpg'):
            name = re.findall(r'[0-9]+', filename)
            filepath_downloaded.append(int(name[0]))

filepath_downloaded = list(set(filepath_downloaded)) # drop Deduplication
filepath_downloaded

[1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100]

In [5]:
data = pd.read_csv(csv_dir)
print('total point:', len(data))
data.head(3)

total point: 597


full_id     osm_id osm_type  osm_versio               osm_timest  osm_uid  \
0  w5668986  5668986.0      way        17.0  2022-09-24T23:47:51.000  88339.0   
1  w5668986  5668986.0      way        17.0  2022-09-24T23:47:51.000  88339.0   
2  w5668986  5668986.0      way        17.0  2022-09-24T23:47:51.000  88339.0   

  osm_user   osm_change  wall  room  ...    tiger_coun  tiger_cfcc  surface  \
0  osmkarl  126599700.0   NaN   NaN  ...  New York, NY         A41  asphalt   
1  osmkarl  126599700.0   NaN   NaN  ...  New York, NY         A41  asphalt   
2  osmkarl  126599700.0   NaN   NaN  ...  New York, NY         A41  asphalt   

              name      highway  distance       angle          x          y  \
0  La Salle Street  residential       0.0  298.966477 -73.956823  40.812622   
1  La Salle Street  residential     100.0  298.946141 -73.957610  40.812952   
2  La Salle Street  residential     200.0  298.885915 -73.958396  40.813283   

   SVI_ID  
0       1  
1       2  
2       3  

[3 rows x 135 columns]

In [6]:
data_not_downloaded = data[~ data.SVI_ID.isin(filepath_downloaded)] # drop downloaded images
data2 = data_not_downloaded.reset_index(drop=True)

print('total image you need to download:',len(data))
print('have downloaded:', len(filepath_downloaded))
print( 'have not downloaded:',len(data2))

data2.head(5)

total image you need to download: 597
have downloaded: 100
have not downloaded: 497


full_id     osm_id osm_type  osm_versio               osm_timest  \
0  w5672571  5672571.0      way        19.0  2022-08-19T02:40:33.000   
1  w5672571  5672571.0      way        19.0  2022-08-19T02:40:33.000   
2  w5672650  5672650.0      way        19.0  2021-10-06T04:11:18.000   
3  w5672650  5672650.0      way        19.0  2021-10-06T04:11:18.000   
4  w5672656  5672656.0      way        21.0  2022-02-23T19:05:07.000   

      osm_uid osm_user   osm_change  wall  room  ...    tiger_coun  \
0    384667.0   MxxCon  125085921.0   NaN   NaN  ...  New York, NY   
1    384667.0   MxxCon  125085921.0   NaN   NaN  ...  New York, NY   
2   2591427.0  snake21  112148304.0   NaN   NaN  ...  New York, NY   
3   2591427.0  snake21  112148304.0   NaN   NaN  ...  New York, NY   
4  14830788.0   emelye  117783589.0   NaN   NaN  ...  New York, NY   

   tiger_cfcc  surface               name      highway  distance       angle  \
0         A41  asphalt  West 122nd Street  residential     200.0  299.056735   
1         A41  asphalt  West 122nd Street  residential     300.0  299.056735   
2         A41  asphalt  West 110th Street    secondary       0.0  140.339708   
3         A41  asphalt  West 110th Street    secondary     100.0  119.163485   
4         A41  asphalt  West 114th Street  residential       0.0  118.602719   

           x          y  SVI_ID  
0 -73.959838  40.811298     101  
1 -73.960623  40.811629     102  
2 -73.959650  40.801156     103  
3 -73.958914  40.800780     104  
4 -73.966962  40.807599     105  

[5 rows x 135 columns]

##  downlaoding images

In [7]:
for i in range(0,497):
# for i in range(len(data2)):
    SVI_ID = data2.loc[i, 'SVI_ID']  # FID
    longitude = data2.loc[i, 'x']  # Lng
    latitude = data2.loc[i, 'y']  # Lat

    heading = data2.loc[i, 'angle']
    # heading =  data2.loc[i,'angle'] - 90  #building elevation
    image_save_dir = image_dir  + '//'+ str(SVI_ID) + '.jpg'

    url = 'https://maps.googleapis.com/maps/api/streetview?size='+ size + '&location=' + str(latitude) + ',' + str(longitude) +  '&key='+  key +  '&fov=' +  str(fov) + '&pitch=' + str(pitch) +  '&heading=' + str(heading) + '&source=outdoor'

    # url = "https://maps.googleapis.com/maps/api/streetview?size=" + size + "&location=" + str(latitude) + "," + str(longitude) + "&heading=" + str(heading) + "&pitch=" + pitch + "&key=" + key + "&fov=" + fov + "&signature=" + YOUR_SIGNATURE
    download(url, image_save_dir, SVI_ID )



image save 101
image save 102
image save 103
image save 104
image save 105
image save 106
image save 107
image save 108
image save 109
image save 110
image save 111
image save 112
image save 113
image save 114
image save 115
image save 116
image save 117
image save 118
image save 119
image save 120
image save 121
image save 122
image save 123
image save 124
image save 125
image save 126
image save 127
image save 128
image save 129
image save 130
image save 131
image save 132
image save 133
image save 134
image save 135
image save 136
image save 137
image save 138
image save 139
image save 140
image save 141
image save 142
image save 143
image save 144
image save 145
image save 146
image save 147
image save 148
image save 149
image save 150
image save 151
image save 152
image save 153
image save 154
image save 155
image save 156
image save 157
image save 158
image save 159
image save 160
image save 161
image save 162
image save 163
image save 164
image save 165
image save 166
image save

# 3.4 Look into the syntax of Google SVI API

    https://maps.googleapis.com/maps/api/streetview?
    size=800x400 
    & location=52.60367961,13.42125453 
    & key=xxxxxxxxx 
    & fov=120 
    & pitch=0 
    & heading=184.95 
    & source=outdoor 

In [ ]:
# the http address that actually calls the API with settings
url